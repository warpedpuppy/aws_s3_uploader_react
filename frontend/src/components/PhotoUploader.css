form {
    width: 700px;
    margin: 150px auto;
}

form div {
    margin: 30px auto;
}

.custom-file {
    position: relative;
    display: inline-block;
    width: 100%;
    height: calc(1.5em + .75rem + 2px);
    margin-bottom: 0;
}
.custom-file-input {
    position: relative;
    z-index: 2;
    width: 100%;
    height: calc(1.5em + .75rem + 2px);
    margin: 0;
    opacity: 0;
}
.custom-file-label {
    position: absolute;
    top: 0;
    right: 0;
    left: 0;
    z-index: 1;
    height: calc(1.5em + .75rem + 2px);
    padding: .375rem .75rem;
    font-weight: 400;
    line-height: 1.5;
    color: #495057;
    background-color: #fff;
    border: 1px solid #ced4da;
    border-radius: .25rem;
}
.custom-file-input:lang(en)~.custom-file-label::after {
    content: "Browse";
    cursor: pointer;
}
.custom-file-label::after {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    z-index: 3;
    display: block;
    height: calc(1.5em + .75rem);
    padding: .375rem .75rem;
    line-height: 1.5;
    color: #495057;
    content: "Browse";
    background-color: #e9ecef;
    border-left: inherit;
    border-radius: 0 .25rem .25rem 0;
}

button[type=submit] {
    padding: 10px 20px;
    border-radius: 5px;
}
.hide {
    display:none;
}
.show {
    display: block;
}
#preview {
    margin: 0 auto;
}
.photo-message-success {
    color: green;
    font-weight: bold;
}
.photo-message-error {
    color: red;
    font-style: italic;
}