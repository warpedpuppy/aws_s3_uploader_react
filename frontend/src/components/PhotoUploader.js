import React from 'react';
import './PhotoUploader.css';
import UploadService from '../services/uploader-service';

export default class PhotoUploader extends React.Component {

    constructor(props){
        super(props)
        this.img = new Image();
        this.img.onload = this.imgLoadHandler;
        this.imgValue = '';
        this.file = null;
        this.requiredSize = {width: 640, height: 480}
        this.state = {
            hideForm: false,
            photoSizeCheck: false,
            photoMessage: ``,
            showImage: false
        }
    }
    
    imgLoadHandler = () => {
            URL.revokeObjectURL(this.file)
            if (this.img.width === this.requiredSize.width && this.img.height === this.requiredSize.height) {
                this.setState({
                    photoMessage: "correct size",
                    photoSizeCheck: true
                })
                let str = this.imgValue;
                document.getElementById('loader-label').innerHTML = str.substring(str.lastIndexOf(`\\`) + 1);
            } else {
                this.setState({
                    photoMessage: `wrong size -- must be ${this.requiredSize.width}x${this.requiredSize.height}`,
                    photoSizeCheck: false
                })
                document.getElementById('event_image').value = null;
                document.getElementById('loader-label').innerHTML = "upload a new image";
            }
    }


    onSubmitHandler = async (e) => {

        if(e) e.preventDefault();
        if (!this.file) {
            this.setState({photoMessage: `please choose an image`});
            return;
        }
        this.setState({hideForm: true})
        if (!this.state.photoSizeCheck) return;

        let res = await UploadService.initUpload('event_image')
        
        if (res) {
            this.photoLoadComplete();
        }
    }


    photoLoadComplete = () => {
        this.setState({photoMessage: "photo uploaded", photoSizeCheck: true, hideForm: false, showImage: true})
        document.getElementById('preview').src = this.img.src;
    }
    

    onChangeHandler = (e) => {
            const { files } = document.getElementById('event_image');
            this.file = files[0];
            var url = URL.createObjectURL(this.file);
            this.img.src = url;
            this.imgValue = e.currentTarget.value;
    }

    render() {
        let photoMessageClass = (this.state.photoSizeCheck)? 'photo-message-success' : 'photo-message-error';
        return (
            <>
                <h1 className={ (!this.state.hideForm) ? 'hide' : 'show' }>loading. . . </h1>
                <form onSubmit={this.onSubmitHandler} className={(this.state.hideForm) ? 'hide' : 'show'}>
                    <div className="file-input-container">
                        <input 
                            type="file" 
                            className="file-input" 
                            name="event_image" 
                            id="event_image"
                            onChange={ this.onChangeHandler }
                            accept=".jpg, .png, .gif, .jpeg" />
                        <label 
                            className="file-label" 
                            id="loader-label" 
                            htmlFor="event_image">
                            click here to choose a new image -- must be 640x480
                        </label>
                        <div id="result" className={photoMessageClass}>{this.state.photoMessage}</div>
                    </div>
                    <div>
                        <button className="form-submit" type="submit">submit</button>
                    </div>
                </form>
                <img src="" id="preview" className={ this.state.showImage ? 'show' : 'hide' } alt="uploaded item" />
            </>
        );
    }
}

