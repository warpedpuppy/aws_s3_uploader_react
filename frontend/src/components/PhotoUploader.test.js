import React from 'react';
import ReactDOM from 'react-dom';
import PhotoUploader from './PhotoUploader';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

describe( 'smoke tests', () => {
  it('photo uploader renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<PhotoUploader />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
})

describe('snapshot tests', () => {
    it('renders the UI as expected', () => {
      const wrapper = shallow(<PhotoUploader />)
      expect(toJson(wrapper)).toMatchSnapshot()
    });
})

describe( 'uploader submit button event listener tests', () => {
    it(`responds with 'please choose an image'`, () => {
      const wrapper = shallow(<PhotoUploader />)
      expect(wrapper.find('div#result').text()).toEqual('')
      wrapper.find(`form`).simulate('submit');
      expect(wrapper.find('div#result').text()).toEqual('please choose an image')
    })
  })