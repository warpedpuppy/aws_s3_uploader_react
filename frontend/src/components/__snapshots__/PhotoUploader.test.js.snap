// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders the UI as expected 1`] = `
<Fragment>
  <h1
    className="hide"
  >
    loading. . . 
  </h1>
  <form
    className="show"
    onSubmit={[Function]}
  >
    <div
      className="file-input-container"
    >
      <input
        accept=".jpg, .png, .gif, .jpeg"
        className="file-input"
        id="event_image"
        name="event_image"
        onChange={[Function]}
        type="file"
      />
      <label
        className="file-label"
        htmlFor="event_image"
        id="loader-label"
      >
        click here to choose a new image -- must be 640x480
      </label>
      <div
        className="photo-message-error"
        id="result"
      />
    </div>
    <div>
      <button
        className="form-submit"
        type="submit"
      >
        submit
      </button>
    </div>
  </form>
  <img
    alt="uploaded item"
    className="hide"
    id="preview"
    src=""
  />
</Fragment>
`;
